x/virtualgroup/keeper/msg_server.go

```diff
@@ -0,0 +1,390 @@
+package keeper
+
+import (
+	"context"
+
+	sdkerrors "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	gnfdtypes "github.com/bnb-chain/greenfield/types"
+	sptypes "github.com/bnb-chain/greenfield/x/sp/types"
+	"github.com/bnb-chain/greenfield/x/virtualgroup/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) UpdateParams(goCtx context.Context, req *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+	if k.GetAuthority() != req.Authority {
+		return nil, sdkerrors.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", k.GetAuthority(), req.Authority)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if err := k.SetParams(ctx, req.Params); err != nil {
+		return nil, err
+	}
+
+	return &types.MsgUpdateParamsResponse{}, nil
+}
+
+func (k msgServer) CreateGlobalVirtualGroup(goCtx context.Context, req *types.MsgCreateGlobalVirtualGroup) (*types.MsgCreateGlobalVirtualGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	var gvgStatisticsWithinSPs []*types.GVGStatisticsWithinSP
+
+	spOperatorAddr := sdk.MustAccAddressFromHex(req.PrimarySpAddress)
+
+	sp, found := k.spKeeper.GetStorageProviderByOperatorAddr(ctx, spOperatorAddr)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound
+	}
+	// TODO(fynn): Check whether the number of secondary sp in gvg matches the redundancy parameters
+	var secondarySpIds []uint32
+	for _, id := range req.SecondarySpIds {
+		ssp, found := k.spKeeper.GetStorageProvider(ctx, id)
+		if !found {
+			return nil, sdkerrors.Wrapf(sptypes.ErrStorageProviderNotFound, "secondary sp not found, ID: %d", id)
+		}
+		secondarySpIds = append(secondarySpIds, ssp.Id)
+		gvgStatisticsWithinSP := k.GetOrCreateGVGStatisticsWithinSP(ctx, ssp.Id)
+		gvgStatisticsWithinSP.SecondaryCount++
+		gvgStatisticsWithinSPs = append(gvgStatisticsWithinSPs, gvgStatisticsWithinSP)
+	}
+
+	gvgFamily, err := k.GetOrCreateEmptyGVGFamily(ctx, req.FamilyId, sp.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Each family supports only a limited number of GVGS
+	if k.MaxGlobalVirtualGroupNumPerFamily(ctx) < uint32(len(gvgFamily.GlobalVirtualGroupIds)) {
+		return nil, types.ErrLimitationExceed.Wrapf("The gvg number within the family exceeds the limit.")
+	}
+
+	gvgID := k.GenNextGVGID(ctx)
```

Comment:
`GenNextGVGID` have been done twice, and thus may cause chaos.

Reply:
Removed it.